src/Orc.Controls/Orc.Controls.Shared/Helpers/DateTimeParser.cs

```diff
@@ -0,0 +1,261 @@
+                    }
+                    else if (formatInfo.MinuteFormat.Length == 2 && partValue.Length != 2) // 'mm'
+                    {
+                        return ThrowOnError<FormatException>("Invalid minute value. Minute must contain 2 digits");
+                    }
+
+                    minute = int.Parse(partValue); // There is no reason to fail.
+
+                    input = input.Substring(partValue.Length);
+                }
+                else if (i == formatInfo.SecondPosition)
+                {
+                    partValue = new string(input.TakeWhile(x => char.IsDigit(x)).ToArray());
+
+                    if (formatInfo.SecondFormat.Length == 1 && (partValue.Length < 1 || partValue.Length > 2))
+                    {
+                        return ThrowOnError<FormatException>("Invalid second value. Second must contain 1 or 2 digits"); // 's'
+                    }
+                    else if (formatInfo.SecondFormat.Length == 2 && partValue.Length != 2) // 'ss'
+                    {
+                        return ThrowOnError<FormatException>("Invalid second value. Second must contain 2 digits");
+                    }
+
+                    second = int.Parse(partValue); // There is no reason to fail.
+
+                    input = input.Substring(partValue.Length);
+                }
+                else if (i == formatInfo.AmPmPosition)
+                {
+                    partValue = new string(input.Take(formatInfo.AmPmFormat.Length).ToArray());
+
+                    if (formatInfo.IsAmPmShortFormat == true && !(Meridiems.IsShortAm(partValue) || Meridiems.IsShortPm(partValue)))
+                    {
+                        return ThrowOnError<FormatException>("Invalid AM/PM designator value");
+                    }
+                    else if (formatInfo.IsAmPmShortFormat == false && !(Meridiems.IsLongAm(partValue) || Meridiems.IsLongPm(partValue)))
+                    {
+                        return ThrowOnError<FormatException>("Invalid AM/PM designator value");
+                    }
+
+                    amPm = partValue;
+
+                    hour += Meridiems.IsPm(amPm) ? 12 : 0;
+
+                    input = input.Substring(partValue.Length);
+                }
+            }
+
+            separator = formatInfo.GetSeparator(i);
+            if (separator != null && separator.Length > 0)
+            {
+                if (!input.StartsWith(separator))
+                {
+                    return ThrowOnError<FormatException>("Invalid value. Value does not match to format");
+                }
+            }
+
+            try
+            {
+                return new DateTime(year, month, day, hour, minute, second);
+            }
+            catch (Exception exc)
+            {
+                return ThrowOnError<FormatException>("Invalid value. It's not possible to build new DateTime object", exc);
```

Comment:
I'm using this ThrowOnError<TException> method to support TryParse and Parse methods like in standard .NET parsers. Throwing exceptions is expensive, that's why I throw or return false depending on throwOnError parameter. Should I add logging to ThrowOnError<TException> method ?